iup/__init__.py

```diff
@@ -0,0 +1,358 @@
+                .sort("date")
+                .filter(pl.col("date") >= start_date)
+            )
+
+        out = CumulativeUptakeData(out)
+
+        return out
+
+
+def get_nis(
+    data_path,
+    region_col,
+    date_col,
+    estimate_col,
+    rollout,
+    filters=None,
+) -> CumulativeUptakeData:
+    out = (
+        (pl.read_csv(data_path))
+        .with_columns(estimate=pl.col(estimate_col).cast(pl.Float64, strict=False))
+        .drop_nulls(subset=["estimate"])
+    )
+
+    if filters is not None:
+        filter_expr = pl.lit(True)
+        for k, v in filters.items():
+            filter_expr &= pl.col(k).is_in(pl.lit(v))
+        out = out.filter(filter_expr)
+
+    out = (
+        out.with_columns(
+            region=pl.col(region_col),
+            date=pl.col(date_col).str.to_date(
+                "%m/%d/%Y %H:%M"
+            ),  # "%m/%d/%Y %I:%M:%S %p"
+        )
+        .select(["region", "date", "estimate"])
+        .sort("region", "date")
+    )
+
+    unique_regions = out["region"].unique()
+    rollout_rows = pl.DataFrame(
+        {
+            "region": unique_regions,
+            "date": [rollout for _ in range(len(unique_regions))],
+            "estimate": [0.0 for _ in range(len(unique_regions))],
+        }
+    )
+    out = out.vstack(rollout_rows).unique(maintain_order=True).sort("date")
+
+    out = CumulativeUptakeData(out)  # rollout.str.to_datetime("%Y-%m-%d"))
+
+    return out
+
+
+def standardize(x, mn=None, sd=None):
+    if mn is not None:
+        return (x - mn) / sd
+    else:
+        return (x - x.mean()) / x.std()
+
+
+def unstandardize(x, mn, sd):
+    return x * sd + mn
+
+
+class UptakeModel(abc.ABC):
+    @abc.abstractmethod
+    def fit(self, data: UptakeData) -> Self:
+        pass
+
+    @abc.abstractmethod
+    def predict(self, data: UptakeData, *args, **kwargs) -> UptakeData:
+        pass
+
+
+class LinearIncidentUptakeModel(UptakeModel):
+    def __init__(self):
+        self.model = LinearRegression()
+
+    def fit(self, data: IncidentUptakeData) -> Self:
+        self.start = data.group_by("region").agg(
+            [
+                pl.col("estimate").last().alias("last_estimate"),
+                pl.col("date").last().alias("last_date"),
+                pl.col("elapsed").last().alias("last_elapsed"),
+                (pl.col("estimate") * pl.col("interval"))
+                .filter(pl.col("season") == pl.col("season").max())
+                .sum()
+                .alias("last_cumulative"),
+            ]
+        )
+
+        data = IncidentUptakeData(  # Not sure why I must reinforce type here
+            data.trim_outlier_intervals().with_columns(
+                previous_std=pl.col("previous").pipe(standardize),
+                elapsed_std=pl.col("elapsed").pipe(standardize),
+                estimate_std=pl.col("estimate").pipe(standardize),
+            )
+        )
+
+        self.previous_mean = data["previous"].mean()
```

Comment:
I think there's some clever way you could do this:
```python
self.standards = [
  getattr(data[x], y)()
  for x in ['previous', 'elapsed', 'estimate']
  for y in ['mean', 'std'])
]
```
although presumably you'd want that as some kind of dictionary